### DISCUSSION TITLE
Extracting node positions from labeled frames - pre training/inference

### DISCUSSION THREAD

**Discussion by anaeguiguren** [talmolab/sleap#1636] [Help!]
Hi!

I wanted to ask if there was a way of accessing the node positions for each instance in each frame from the .slp data (or .h5)?
I am currently working on a project analyzing the orientation and spatial spread of sperm whales from drone footage. So far, I have had really nice results from just 200 frames labeled, with varying flight altitudes (hence, individual sizes) and lighting structures. It's even doing a good job at identifying whales that are slightly below the surface!
However, As I label more frames, I would like to conduct some analyses on the resulting positions from manually labelled instances. I followed the instructions in the https://github.com/talmolab/sleap/blob/develop/docs/notebooks/Data_structures.ipynb notebook, and can see the location for each node in each instance and each frame by looking at the _labeled_frame.instances_ 

But I'm not sure if there's a way incorporated into the program to export this data as a csv?

Thanks so much in advance for your time, and for creating this software and making it so user-friendly despite it being a pretty complex system! :)

**Comment by Marielamend** [comment]
Hello, @anaeguiguren!,

In regards to exporting your data as a csv, you can go ahead and go to **File** in SLEAP and select (**Export Analysis CSV...**).
![Screenshot 2023-12-13 125845](https://github.com/talmolab/sleap/assets/120222245/da9a0002-7709-40e5-8e4a-a85965833778)

You can also use [this notebook as a reference](https://github.com/talmolab/ws-remoras/blob/main/analysis.ipynb).

Let us know if you have any other questions!
Best, 
Mariela!

**Reply by anaeguiguren** [reply]
Hi Mariela, Thanks so much for your reply! I've tried this method, and it results in a table like the one below, in which there is only one position for each node in each frame, despite there being 19 labelled instances on that same frame:
![Screenshot 2023-12-15 104112](https://github.com/talmolab/sleap/assets/34753317/471f2884-7599-4a3c-a11b-c2400999ffe5)
![Screenshot 2023-12-15 104143](https://github.com/talmolab/sleap/assets/34753317/b8ef0885-fbb2-4ab0-a888-6ef812d4f6d8)

Is there something I can do so that all the instances show up? Thanks so much again for your time!

Best, 
Ana

**Comment by Marielamend** [comment]
Hello @anaeguiguren!

Once you are getting accurate frame-by-fame predictions, just as you mentioned you are! You now want to run predictions for the entire video clips so you can correct the instances across frames.

**1.** You can select "**Run Inference...**" under **Predict**.
<img width="294" alt="Screen Shot 2023-12-18 at 11 36 48 AM" src="https://github.com/talmolab/sleap/assets/120222245/4a1b3f80-f5a1-4128-9eb9-0ca2458483e4">

**2.** Then select a tracker method:
<img width="377" alt="Screen Shot 2023-12-18 at 11 42 44 AM" src="https://github.com/talmolab/sleap/assets/120222245/cd0ac1a3-cc64-498e-8891-fafb37fadc70">

You can leave the default settings for the similarity and matching method. If you’d like to get further explanation on the different methods for tracking you can reference this [part of the documentation.](https://sleap.ai/guides/proofreading.html#tracking-method-details.)
<img width="578" alt="Screen Shot 2023-12-18 at 12 13 47 PM" src="https://github.com/talmolab/sleap/assets/120222245/2e6fb8a7-d831-49e3-a6d6-679ef8d8b3ff">



**3.** Lastly, you want to proofread once you have your predicted tracks, to make sure to correct any mistakes. The two most common mistakes are Mistaken identities and Lost identities. To correct these:

   **a.) Mistaken identities**: If there is an identity swap you can transpose a pair of instances by selecting (Ctrl+1-9) to swap them back in the correct order. Or you can go under **Tracks** and select "**Set Instance Track**" and use the drop menu to select the correct track.
    
   **b.) Lost identities**: If there is an instance that was lost you can use the shortcut (Ctrl+e) to go to  “Next Track Spawn Frame” to go to the last frame where the new track showed to correct the track from there.



You can reference your Keyboard shortcuts under **Help**.

![image](https://github.com/talmolab/sleap/assets/120222245/00c46610-d563-4736-8cf9-2f61f9a30323)


**4.** Once you are done with proofreading, you can export your data for analysis as a CSV file!

Anything else let us know!
Best, 
Mariela

**Comment by anaeguiguren** [comment]
Hi Mariela, 
Thanks again for your reply!
I was hoping to find if there was a way of just obtaining the manual label data, before running training and inference?
Either through the user interface or a function that could extract that data from the .slp file? If not, I'll give it a shot to write it up! :) Thanks again for your help!

**Reply by Marielamend** [reply]
Hello @anaeguiguren!

So sorry your comment fell through the cracks. Ah I see what you are saying!, we suggested you to run training because when we export slp files to analysis files we gather instances into tracks before exporting so if there are no tracks in the project (before running predictions), then the export function assumes that the project is a single animal project. 

Let us know how its going and if we can further assist!

Best, 
Mariela
